Manager now handles hierarchy

LizzyRandall opened the pull request:
New class handles collapsing/expanding its piece and the data while the
manager handles collapsing itself (if hierarchy) and updating the
sections locations when one expands/collapses. It also still handles creation of the cells from the tableview passed by the viewcontroller. Now that I'm thinking about it, maybe the section can handle creating the cell by putting the info it has from the data.. anyway, for now the manager is handling that.

I changed the arrows to ones from a github repo based on iOS images: https://github.com/youknowone/UIKitResources. I also renamed them based on the standards. It was black and I changed the hue to gray. I thought it looked nicer and the repo says "This resources are included in proprietary software." so should be fine to use, right?

I made a custom tableview cell subclass to handle defaults on the way a
header and toggle check like cell should look. It contains an enum for
cell types besides headers and checkmark cells.

@michaelharro Please let me know of any feedback for it. I scanned over the standards again before the commit to make sure I caught everything but I probably missed some things. Thanks!

Commit:
Manager now handles hierarchy

New class handles collapsing/expanding its piece and the data while the
manager handles collapsing itself (if hierarchy) and updating the
sections locations when one expands/collapses.

I changed the arrow which is from github repo based on iOS images. It
was black and I changed the hue to gray.

I made a custom tableview cell subclass to handle defaults on the way a
header and toggle check like cell should look. It contains an enum for
cell types besides headers and checkmark cells.

michaelharro commented on CollapsibleExample.xcodeproj/xcuserdata/lizzyrandall.xcuserdatad/xcschemes/CollapsibleExample.xcscheme:
```diff
@@ -83,6 +83,11 @@
          </BuildableReference>
       </BuildableProductRunnable>
       <AdditionalOptions>
+         <AdditionalOption
```
You might want to update your gitignore file so it doesn't include personal settings files. Here is MissionHub's https://github.com/CruGlobal/MissionHub-iOS/blob/master/.gitignore

michaelharro commented on CollapsibleExample/MHCollapsibleSection.h:
```diff
@@ -0,0 +1,63 @@
+//returns result of if row number is in range for section
+- (BOOL)rowNumInRange:(NSUInteger)num;
+
+//returns label at a specified row, note: this is offset since the labels themselves
```
Great naming on all these methods. I also like your commenting. A helpful tool that will make your commenting quicker and format your comments according to apple's standards is VVDocumenter. If you install alcatraz (an xcode plugin manager, http://alcatraz.io/ ) and search for it I think you will enjoy it. You should also look around alcatraz for plugins that make your life easier.

michaelharro commented on CollapsibleExample/MHTableViewCell.h:
```diff
@@ -0,0 +1,51 @@
+//calls super first then sets defaults
+- (instancetype) initWithStyle:(UITableViewCellStyle)style reuseIdentifier:(NSString *)reuseIdentifier;
+
+- (void)setMHCellDefaultsWithAccessory:(UITableViewCellAccessoryType)accessory style:(UITableViewCellSelectionStyle)style labelColor:(UIColor*)color accessoryView:(UIView*)view;
```
I am loving your naming. It is on point. This is the only thing I could pick up and it is a confusing part of objc programming. Normally you wouldn't use MH prefix in your code. The prefix normally only occurs in naming a class. When referencing an instance of a class with the prefix you don't use it. So in this case the name would be:

```
- (void)setCellDefaultsWithAccessory:(UITableViewCellAccessoryType)accessory style:(UITableViewCellSelectionStyle)style labelColor:(UIColor*)color accessoryView:(UIView*)view;
```

It is a confusing way to do namespacing. Fortunately they don't use prefixes in swift, they actually implemented a proper namespacing system.

michaelharro commented on FilterViewController.m:
```diff
@@ -22,20 +24,56 @@ @implementation FilterViewController
+    MHCollapsibleViewManager *interactions = [[MHCollapsibleViewManager alloc] initManagerWithAnimation:UITableViewRowAnimationMiddle topHierarchyTitle:@"Interactions" tableView:self.tableView];
     //Index 0 is title
-    [assignTo  initializeData:[NSMutableArray arrayWithObjects:@"Jan", @"Andy", @"Peggy", @"Maggie", nil]];
+    [interactions setDataWithFilterNames:[[NSArray alloc] initWithObjects:[self returnInteractionsArray],nil] headerTitles:[[NSArray alloc] initWithObjects:@"Interactions", nil]];
```
This also a very well written class. This is a minor thing. You can write this more concisely with an array literal (literals are the things with @ in front of them. They expand to what you wrote.):

```
[interactions setDataWithFilterNames:@[[self returnInteractionsArray]] headerTitles:@[@"Interactions"]];
```

You could simlify it further using the dot notation on the returnInteractionsArray and a shorter name for the method

```
[interactions setFilterNames:@[self.returnInteractionsArray] headerTitles:@[@"Interactions"]];
```

Note: On the topic of literals, if you see a string with an @ in front it will also expand to [[NSString alloc] initWith...] which is helpful to know so you know what methods are available on those objects.

michaelharro commented:
@LizzyRandall I am moving my stuff to Orlando this week, we will be in CO doing IBS and CSU for the following 6 weeks, then I will be moving into an apartment the week after that. So I have not been able to do an in depth review. However, here are a few thoughts:
The code is beautifully written, very easy to read. The way you have structured your classes is fantastic and I would like to use it as an example for other iOS devs on our team who are learning. It also works well. Excellent work!

I don't know how much feedback I will be able to give you for the next 2 months. But here is some direction I would like to give for next steps.
1. We already talked about the next step: Getting user input for the 4 different survey input types.
2. We also talked about this: sending data to another view controller via either a delegate or a set of notificaitons
3. We haven't talked about: integrating with MissionHub. You will need to get to know the models https://github.com/CruGlobal/MissionHub-iOS/tree/master/MissionHub/Models (CoreData models are further explained on this pull request https://github.com/CruGlobal/godtools-ios/pull/5 ) you will also need to get to know the request options class https://github.com/CruGlobal/MissionHub-iOS/blob/master/MissionHub/APIControllers/MHRequestOptions.h the people list has an instance of this class, when you refresh the people list it sends a request using all the options listed in that object to the API and then displays the response. So when you return your filters after the apply/save button is hit you should add them to that object and refresh the people list. The final part of integrating your code into the MH app is to have your filter view controller take an array of fields and survey objects and display them in your filter view controller. There are lots of ways to do it, the easiest being to just modify your current class so it is tied to MissionHub. However, if you have time I would like you to try the follow step, I think it will be a great learning exercise for you.
4. Keep your current solution generalized as it currently is. Create a pod out of it https://guides.cocoapods.org/making/making-a-cocoapod.html, add the Pod to MissionHub's Podfile https://github.com/CruGlobal/MissionHub-iOS/blob/master/Podfile, make MissionHub specific subclasses of your classes. I intend to make a private spec repo for CruGlobal when I get back. I'll let you decide if you want to give that a go https://guides.cocoapods.org/making/private-cocoapods.html or just submit it to the master spec repo.

I hope all this makes sense. Let me know if you have any questions. I can't guarantee I'll answer them quickly but I will try to answer them when I am able.

LizzyRandall commented:
I didn't know you were coming to Orlando! I hope you have safe travels and a good time at CSU.

Thanks for the feedback. I'll work towards those bullet points and I like the cocoapods idea for integrating into MH. The manager might be useful for other apps so that sounds like the most usability route for the code.

Commit:
Changes Harro suggested

Changed arrays to be instantiated with literals instead
Changed the naming of two of the methods for MHTableViewCell to not
include MH
Added gitignore, which hopefully will work as I copied the one from MH
and removed a few lines that looked specific to MH

LizzyRandall commented:
Merging pull request since I added the changes suggested. Now onto modals!
